system.go

```diff
@@ -1,12 +1,43 @@
 package nitriding
 
-import "syscall"
+import (
+	"errors"
+	"io"
+	"syscall"
+)
 
 const (
-	defaultFdCur = 65536
-	defaultFdMax = 65536
+	defaultFdCur     = 65536
+	defaultFdMax     = 65536
+	errTooMuchToRead = "reached read limit"
 )
 
+// limitReader behaves like a Reader but it returns errTooMuchToRead if the
+// read limit was met or exceeded.
+type limitReader struct {
+	io.Reader
+	Limit int
+}
+
+func (l *limitReader) Read(p []byte) (int, error) {
+	if l.Limit <= 0 {
+		return 0, errors.New(errTooMuchToRead)
+	}
+	if len(p) > l.Limit {
+		p = p[0:l.Limit]
+	}
+	n, err := l.Reader.Read(p)
+	l.Limit -= n
```

Comment:
Just to check: `int` is big enough to hold the length of any slice, so this can't underflow? The go spec is ambiguous about when `int` is 32 bit, when it's 64 bit, but implies the upper half-range is excluded so you can use a signed index?

Reply:
Yes, `int` should be big enough for that. Go's built-in functions [len](https://pkg.go.dev/builtin#len) and [cap](https://pkg.go.dev/builtin#cap) (which can determine a slice's length and capacity, respectively) both return an `int`. To quote the specification:

> The built-in functions len and cap take arguments of various types and return a result of type int. The implementation guarantees that the result always fits into an int.